o1vm/src/mips/singlestep.rs

```diff
@@ -0,0 +1,116 @@
+// Amalgamation of the different MIPS circuits into a "single" circuit.
+
+use super::column;
+
+pub trait SingleStepper {
+    type S<I, P, V>;
+
+    type Position;
+
+    type Instruction;
+
+    type Variable: Clone
+        + std::ops::Add<Self::Variable, Output = Self::Variable>
+        + std::ops::Sub<Self::Variable, Output = Self::Variable>
+        + std::ops::Mul<Self::Variable, Output = Self::Variable>
+        + std::fmt::Debug
+        + ark_ff::Zero
+        + ark_ff::One;
+
+    type State = S<Self::Instruction, Self::Position, Self::Variable>;
+
+    /// Helpers
+
+    fn alloc_scratch(state: Self::State) -> (Self::Position, Self::State);
+
+    fn variable(state: &Self::State, column: Self::Position) -> Self::Variable;
+
+    fn add_constraint(state: Self::State, assert_equals_zero: Self::Variable) -> Self::State;
+
+    // ABORTS
+    fn check_is_zero(assert_equals_zero: &Self::Variable);
+
+    fn assert_is_zero(state: Self::State, assert_equals_zero: Self::Variable) -> Self::State {
+        Self::check_is_zero(&assert_equals_zero);
+        self.add_constraint(assert_equals_zero)
+    }
+
+    // ABORTS
+    fn check_equal(x: &Self::Variable, y: &Self::Variable);
+
+    fn assert_equal(state: Self::State, x: Self::Variable, y: Self::Variable) -> Self::State {
+        Self::check_equal(&x, &y);
+        self.add_constraint(x - y)
+    }
+
+    // ABORTS
+    fn check_boolean(x: &Self::Variable);
+
+    fn assert_boolean(state: Self::State, x: Self::Variable) {
+        Self::check_boolean(&x);
+        self.add_constraint(x.clone() * x.clone() - x); // polynomial with roots {0, 1}
+    }
+
+    fn add_lookup(state: Self::State, lookup: Lookup<Self::Variable>) -> Self::State;
+
+    // compared to original, this is divided by 4.
+    fn instruction_counter(state: &Self::State) -> Self::Variable;
+
+    fn increase_instruction_counter(state: Self::State) -> Self::State;
+
+    /// Actual stepper
+
+    fn step(state: Self::State, instructions: &Vec<Self::Instruction>) -> Self::State;
+}
+
+/// A simple example
+
+#[derive(
+    Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd, EnumCount, EnumIter
+)]
+pub enum ExampleRegister {
+    A,
+    B,
+    C,
+    D,
+    Lo,
+    Hi,
+    Ip,
+    NextIp,
+}
+
+pub type ExamplePosition = column::ColumnAlias;
+
+pub type ExampleVariable = ark_bn254::Bn254;
```

Comment:
I understand that variables are a way to index to positions in the plonk matrix.
Therfore, I would use something like unsigned integer, but not a crypto related object

Reply:
Yes, I changed to unsigned integer in the latest version.

Regarding the from-scratch approach shown here, this is mostly a playground where I can test out ideas which I will then port back to the regular interpreter, I think, without the full complexity of the regular interpreter.  It didn't start out as that, mind you!

But I think you and Danny are right.